src/core/algorithms/ind/spider/spider.cpp

```diff
@@ -0,0 +1,144 @@
+/** \file
+ * \brief Spider algorithm
+ *
+ * Spider algorithm class methods definition
+ */
+#include "spider.h"
+
+#include <functional>
+#include <memory>
+#include <queue>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+#include "attribute.h"
+#include "config/equal_nulls/option.h"
+#include "config/error/option.h"
+#include "config/mem_limit/option.h"
+#include "config/names_and_descriptions.h"
+#include "config/option_using.h"
+#include "config/thread_number/option.h"
+#include "util/timed_invoke.h"
+
+namespace algos {
+
+using AttributeIndex = spider::AttributeIndex;
+
+Spider::Spider() : INDAlgorithm({}) {
+    DESBORDANTE_OPTION_USING;
+
+    RegisterOption(config::EqualNullsOpt(&is_null_equal_null_));
+    RegisterOption(config::ThreadNumberOpt(&threads_num_));
+    RegisterOption(config::MemLimitMBOpt(&mem_limit_mb_));
+    RegisterOption(config::ErrorOpt(&max_ind_error_));
+    MakeLoadOptsAvailable();
+}
+
+void Spider::MakeLoadOptsAvailable() {
+    MakeOptionsAvailable({config::EqualNullsOpt.GetName(), config::ThreadNumberOpt.GetName(),
+                          config::MemLimitMBOpt.GetName()});
+}
+
+void Spider::MakeExecuteOptsAvailable() {
+    MakeOptionsAvailable({config::ErrorOpt.GetName()});
+}
+
+void Spider::LoadDataInternal() {
+    auto create_domains = [&] {
+        domains_ = model::ColumnDomain::CreateFrom(input_tables_, mem_limit_mb_, threads_num_);
+    };
+    timings_.load = util::TimedInvoke(create_domains);
+}
+
+namespace {
+template <typename Attribute>
+std::vector<Attribute> InitAttributes(std::vector<model::ColumnDomain> const& domains) {
+    std::vector<Attribute> attrs;
+    AttributeIndex attr_count = domains.size();
+    attrs.reserve(attr_count);
+    for (AttributeIndex attr_id = 0; attr_id != attr_count; ++attr_id) {
+        attrs.emplace_back(attr_id, attr_count, domains[attr_id]);
+    }
+    return attrs;
+}
+
+template <typename Attribute>
+std::vector<Attribute> GetProcessedAttributes(std::vector<model::ColumnDomain> const& domains,
+                                              config::EqNullsType is_null_equal_null) {
+    using AttributeRW = std::reference_wrapper<Attribute>;
```

Comment:
I use a `std::reference_wrapper` because I want to create a "vector of references". This class stores a pointer and has a reference casting operator `operator T&()`, as well as a constructor from reference.

Next, I will talk about the advantages of using this class compared to a raw pointer.
We can use a constructor from two iterators (just pass `attrs.begin()`, `attrs.end()`). In the case of raw pointers, we call priority queue constructor, then write loop to push pointers. 
Also, if we used raw pointers, we would have to redefine the compare operator, since we would not be able to use just `std::greater<Attribute>`.

However, we had to write `attr_rw.get()` several times, it seems unavoidable. 

To summarize, I think that in general there are noticeably more advantages, since I managed to avoid writing some code.

I also like the semantic approach to using a `std::reference_wrapper`. When we see raw pointer, then we wonder if it could be a null? In the case of a `std::reference_wrapper`, we work with it as with a reference.